### PR TITLE
Provide PHP 8.1 support

### PR DESCRIPTION
**Pull request by weierophinney**
This patch provides support for PHP 8.1, via the following:

- Adds `~8.1.0` to the list of allowed PHP versions.
- Modifies how the package replaces zend-db:
  - Renames "replace" section of package with "conflict"
  - Changes zend-db constraint in that section to "*"
  - Removes laminas-zendframework-bridge
- Bumps dependencies to versions known to work with PHP 8.1.
- Bumps the laminas-coding-standard version to 2.2 and applies its ruleset.
- Fixes PHPUnit deprecation warnings
- Adds `#[ReturnTypeWillChange]` attributes to methods implementing internal interfaces or extending internal classes.
- Removes calls to `mysqli::init()`, as they were unnecessary and done incorrectly anyways.
- Makes optional arguments required when later arguments are already required.
- Ensures that arguments passed to internal functions are of the correct type (e.g., strings for `str*()` functions, integer exception codes)
- Updates the Vagrantfile, related fixtures, and base `phpunit.xml.dist` to work correctly for local environments.

Replaces #204

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,5 +7,52 @@ oriented API to build the queries. `Laminas\Db` consumes different storage adapt
...
+The repository includes a `Vagrantfile` which allows you to fire up a [vagrant box](https://app.vagrantup.com) with several of our target databases, including:
+
+- MySQL
+- PostgreSQL
+- SQL Server
+
+To start up vagrant:
+
+```bash
+$ vagrant up
```
Comment (Ocramius) [inline README.md:39]: One day, vagrant will be gone for good :D

File: src/Adapter/Driver/IbmDb2/Result.php [hunk 2]
```diff
@@ -176,10 +177,11 @@ public function getFieldCount()
...
 
     /**
-     * @return null|int
+     * @return int
      */
+    #[ReturnTypeWillChange]
     public function count()
     {
-        return;
+        return 0;
```
Comment (Ocramius) [inline src/Adapter/Driver/IbmDb2/Result.php:185]: Whoa, what is this funky business? I know DB2 is a steaming pile of... "legacy", but this seems borked :D

File: src/Adapter/Platform/PlatformInterface.php [hunk 3]
```diff
@@ -56,8 +56,8 @@ public function quoteValue($value);
      *
      * The ability to quote values without notices
      *
-     * @param $value
-     * @return mixed
+     * @param scalar $value
```
Comment (Ocramius) [inline src/Adapter/Platform/PlatformInterface.php:59]: Not really a BC break, but we should avoid these changes in minor releases

File: src/Metadata/Object/ViewObject.php [hunk 4]
```diff
@@ -62,8 +67,9 @@ public function setIsUpdatable($isUpdatable)
         return $this;
     }
 
+    /** @return bool */
     public function isUpdatable()
     {
-        return $this->isUpdatable;
+        return (bool) $this->isUpdatable;
```
Comment (Ocramius) [inline src/Metadata/Object/ViewObject.php:73]: Sneaky change here too, but I guess it's OK from a consumer PoV

**Review by Ocramius — APPROVED** [review]
I went through it - fairly sure that some change **will** lead to regressions, specifically when changing `==` to `===`, but looking at the amount of magic and adapters in the codebase, it is impossible for me to determine where that will happen.

Overall, this change is a good step forward, so let's roll with it as-is, hoping for improved psalm type coverage in the future :)

**Comment by Ocramius** [comment]
Thanks for this huge chunk of work, @boesing and @weierophinney!

**Comment by weierophinney** [comment]
> fairly sure that some change **will** lead to regressions, specifically when changing `==` to `===`,

I was pretty careful about these, and only did strict checks where the type on both sides was essentially known. If I had any doubts, I threw in a `phpcs:ignore` note. (I actually ran into a few cases early on when testing where the strict checks broke things, and at that point got far more picky about when I would make the change!)

But absolutely - there's so much weird stuff with databases, who knows what's going to break. :shrug: :upside_down_face:

File: src/Adapter/Driver/IbmDb2/Result.php (hunk 2 above)
Comment (weierophinney) [reply src/Adapter/Driver/IbmDb2/Result.php:185]: I'm a bit shocked we were not returning AN INTEGER from a `count()` method as well!

File: src/Adapter/Platform/PlatformInterface.php (hunk 3 above)
Comment (weierophinney) [reply src/Adapter/Platform/PlatformInterface.php:59]: The way it was written before, it was essentially missing, though. Should I have said `mixed` instead? I was trying to make these align to what the code actually expects...

File: src/Metadata/Object/ViewObject.php (hunk 4 above)
Comment (weierophinney) [reply src/Metadata/Object/ViewObject.php:73]: Consumer expects a boolean, so :shrug:

**Comment by weierophinney** [comment]
Thanks for reviewing, @Ocramius - I hope never to have a patch quite this big again!

**Pull request merged**